### ISSUE TITLE
Lag spike when crossing certain chunk borders

### ISSUE THREAD

**Issue by filippvl** [PaperMC/Paper#5089]
### Description of issue:
I've encountered some sort of lag spike which I narrowed down to be a floating structure. This structure can be a single block.

The lag spike seems to occur when I'm moving into the spawn chunks while an elevated block is placed on the opposite side of the spawn chunks. It does not occur in single player and occurs when the block is at Y96 or higher.

I've copied this description from: https://bugs.mojang.com/browse/MC-162253
It's the same issue, only certain borders lag spike. Can't find the floating structure mentioned.

### Plugin list:
[11:29:58 INFO]: Plugins (8): AuthMe, ClearLag, CoreProtect, Essentials, EssentialsChat, EssentialsProtect, EssentialsSpawn, Vault

List of VanillaTweaks that I use on the server:
https://pastebin.com/Fp2Dzy0M

### bukkit.yml, spigot.yml, paper.yml, server.properties
bukkit.yml https://pastebin.com/yraYHyGB
spigot.yml https://pastebin.com/PM6WEuYC
paper.yml https://pastebin.com/r84Vy7Jg
server.properties https://pastebin.com/cwHgycBn

### Other helpful links
https://bugs.mojang.com/browse/MC-162253

here's a video of the lag spikes:
https://streamable.com/y7qt9i

### Paper version:
[11:28:48 INFO]: This server is running Paper version git-Paper-416 (MC: 1.16.4) (Implementing API version 1.16.4-R0.1-SNAPSHOT)
[11:28:48 INFO]: Previous version: git-Paper-339 (MC: 1.16.4)
[11:28:48 INFO]: You are running the latest version

**Comment by Chew** [comment]
Continuation of #3468

Also, please update

**Comment by filippvl** [comment]
I've updated the server, problem still exists.

[09:11:47 INFO]: This server is running Paper version git-Paper-439 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)
[09:11:47 INFO]: Checking version, please wait...
[09:11:47 INFO]: Previous version: git-Paper-416 (MC: 1.16.4)
[09:11:47 INFO]: You are running the latest version

**Comment by Fejm** [comment]
```java
	@EventHandler(priority = EventPriority.LOWEST)
	public void onChunkLoadEvent(ChunkLoadEvent event)
	{
		for (int x = 0; x < 16; x++)
		{
			for (int z = 0; z < 16; z++)
			{
				event.getChunk().getBlock(x, 255, z).setType(Material.BARRIER, false);
			}
		}
	}
```

As temporary fix. I wonder if we can do this on packet side.

**Comment by Eptiiix** [comment]
> 
> 
> ```java
> 	@EventHandler(priority = EventPriority.LOWEST)
> 	public void onChunkLoadEvent(ChunkLoadEvent event)
> 	{
> 		for (int x = 0; x < 16; x++)
> 		{
> 			for (int z = 0; z < 16; z++)
> 			{
> 				event.getChunk().getBlock(x, 255, z).setType(Material.BARRIER, false);
> 			}
> 		}
> 	}
> ```
> 
> As temporary fix. I wonder if we can do this on packet side.

Didn't fix it, just a big plate in sky as barriers, but lag spikes when entering another chunk are still there.
Maybe important info: my world was upgraded from 1.8 to 1.13 > 1.14 > 1.15 > 1.16
But there is definitly a context between crossing chunks and flying blocks. (tested)

**Comment by Lypt0x** [comment]
By entering a chunk border where the lag spike occurs,
type **/setworldspawn**. After this command, for that chunk the lag spike is suddenly gone.
You should keep a look at this command to build a workaround to each chunk border.

**Comment by MinecraftCitiesNetwork** [comment]
> By entering a chunk border where the lag spike occurs,
> type **/setworldspawn**. After this command, for that chunk the lag spike is suddenly gone.
> You should keep a look at this command to build a workaround to each chunk border.

Sadly doesn't work for me, I wonder why this worked?

**Comment by Eptiiix** [comment]
> > By entering a chunk border where the lag spike occurs,
> > type **/setworldspawn**. After this command, for that chunk the lag spike is suddenly gone.
> > You should keep a look at this command to build a workaround to each chunk border.
> 
> Sadly doesn't work for me, I wonder why this worked?

i think ther're differnt types of chunk lags in minecraft right now.
On the Mojang Bugs Report Dashboard there is also a chunk lag which was fixed, but my chunk lag still exists.

**Comment by MinecraftCitiesNetwork** [comment]
> > > By entering a chunk border where the lag spike occurs,
> > > type **/setworldspawn**. After this command, for that chunk the lag spike is suddenly gone.
> > > You should keep a look at this command to build a workaround to each chunk border.
> > 
> > 
> > Sadly doesn't work for me, I wonder why this worked?
> 
> i think ther're differnt types of chunk lags in minecraft right now.
> On the Mojang Bugs Report Dashboard there is also a chunk lag which was fixed, but my chunk lag still exists.

Ah right, I really hope this can be fixed soon!

**Comment by Brokkonaut** [comment]
Just to mention: This bug still exists in 1.17

**Comment by Brokkonaut** [comment]
I just tested the patch mentioned here: https://github.com/PaperMC/Paper/issues/5785#issuecomment-861693228 and it definitely helps against the heavy client lag spikes on my skyblock server. So I would recommend to readd it in some way.

**Comment by electronicboy** [comment]
We know that the bug still exists, we had many reports saying that it didn't help and some saying that it did, as well as people upset because it lovingly caused more network load, #PlzMojang

**Comment by Brokkonaut** [comment]
Ok, with that downside.. maybe making it configurable would be an option. I just can say that it really helps against lag spikes on my server. Oh and one thing I forgot in the last comment: There is an issue in that patch in line 45, it should be "return new BitSet();" or it will NPE with completely empty chunks.

**Comment by kennytv** [comment]
I updated the paste to fix that, thanks. I also have to note I haven't actually tested if the ceiling mask method works the same as before, and the previous implementation was *really* magic and hacky.

**Comment by Brokkonaut** [comment]
I did some changes with the patch, because I think it still had some bugs. If anyone wants to test: https://cpaste.de/kukezerecu.diff
I tested it on my server and now could no longer see any of those chunk load lag spikes.

Things I changed:
- The BitMask calculation was incorrect as far as I can see
- When comparing the BitMasks, the own chunks BitMask was modified so it did only work correctly in one direction and might have failed in the other directions
- I removed the second loop and the change of the Packet array, because that was not needed. Now everything is in one loop and it is much simpler.

TODO:
- In the patch "No-Tick view distance implementation" the private bit mask calculation methods (for the old fixed world limits) were added, but those methods are never used anywhere and won't even work in 1.17. They are currently overwritten by this patch with the BitMask methods, but I think those changes should be removed from the other patch, because they are useless.

**Comment by Eptiiix** [comment]
has the patch been removed?
the lag spikes are back since tuinity got merged

**Comment by Brokkonaut** [comment]
The patch was never added to paper 1.17. If you want to build paper yourself, this is the  current version of the patch that we use on our server: https://gist.github.com/Brokkonaut/eb06809a7ec4cab0a581b7d2e34f70e8

**Comment by ghost** [comment]
I'm having this EXACT same issue. I'm running a Paper 1.17.1 server and every time I import a map the map has huge lag/FPS drops only when crossing borders in specific chunks. Doesn't seem to be happening anywhere else on the map. And as far as I know, no one notices these lag spikes.

**Some things I learned:**
- Doesn't affect my Bedrock Minecraft account
- RAM doesn't make a difference
- Setting world spawn in the weird chunks doesn't work
- Optifine doesn't make a difference
- Only happens when entering a new chunk
- Only seems to be affecting worlds/maps I import, and also when I copy and paste those imports into another world

**THE ONLY (PARTIAL) SOLUTION:**
Lower your render distance to 2 and it'll be negligible. You could even up it to 5 render distance, possibly. However, in some chunks the render distance doesn't matter. Which sucks; this sucks...

**Comment by Kryniowesegryderiusz** [comment]
Still happens on 1.18.2 :c

**Comment by kennytv** [comment]
Should be fixed by https://github.com/PaperMC/Paper/commit/0bf8790c9ccd7e7e694bb61839e70948359ee53c

**Comment by molor** [comment]
_(I'll hug you if u port that fix to 1.18.2)_
